File: core/src/utils/rdfa-extractor.ts
// core/src/utils/rdfa-extractor.ts
//
// Parses HTML via parse5 and extracts RDFa items (`[typeof]`
// elements inside a `[vocab]` scope) into JSON-LD-shaped objects
// that the existing per-item validator can consume.
//
// Vocabulary resolution is top-down: as we recurse, we track the
// "current vocab" inherited from ancestors. When we encounter a
// `[typeof]` element, we use the inherited vocab to resolve a bare
// type name. If `typeof` is a fully-qualified schema.org URL, no
// vocab is needed.
//
// Per-element value-extraction rules (RDFa):
//
//   element has `content=` attribute  → content value (RDFa override)
//   meta                              → content
//   a, link                           → href
//   img/audio/video/...               → src
//   object                            → data
//   data, meter                       → value
//   time                              → datetime (fallback to text)
//   element with typeof=              → recursive extraction (nested obj)
//   element with resource= (no typeof) → resource value (URI ref)
//   anything else                     → trimmed textContent
//
// Cycle-7 limitations (documented):
//   - CURIE prefixes (typeof="schema:Product") not supported.
//   - Non-schema.org vocab subtrees silently produce no items.

import { parse } from "parse5";
import type { Issue } from "../types.js";
import { loadRegistry } from "../registry.js";
import { noVocab, invalidItemtype } from "../errors.js";

const registry = loadRegistry();

interface Attribute {
  name: string;
  value: string;
}
interface ParseElement {
  tagName: string;
  attrs: Attribute[];
  childNodes: ParseNode[];
}
interface ParseText {
  nodeName: "#text";
  value: string;
}
type ParseNode = ParseElement | ParseText | { childNodes?: ParseNode[] };

function isElement(node: unknown): node is ParseElement {
  return (
    typeof node === "object" &&
    node !== null &&
    typeof (node as ParseElement).tagName === "string" &&
    Array.isArray((node as ParseElement).attrs)
  );
}

function isTextNode(node: unknown): node is ParseText {
  return (
    typeof node === "object" &&
    node !== null &&
    (node as ParseText).nodeName === "#text" &&
    typeof (node as ParseText).value === "string"
  );
}

function getAttr(el: ParseElement, name: string): string | undefined {
  for (const attr of el.attrs) {
    if (attr.name === name) return attr.value;
  }
  return undefined;
}

function hasAttr(el: ParseElement, name: string): boolean {
  return getAttr(el, name) !== undefined;
}

function children(node: ParseNode): ParseNode[] {
  return (node as { childNodes?: ParseNode[] }).childNodes ?? [];
}

function textContent(node: ParseNode): string {
  if (isTextNode(node)) return node.value;
  if (isElement(node)) {
    return node.childNodes.map(textContent).join("");
  }
  return children(node).map(textContent).join("");
}

const SCHEMA_VOCAB_RE = /^https?:\/\/(?:www\.)?schema\.org\/?$/;
const SCHEMA_TYPE_URL_RE =
  /^https?:\/\/(?:www\.)?schema\.org\/([A-Za-z][A-Za-z0-9]*)\/?$/;

/** Extract a value from a property element per RDFa rules. */
function extractElementValue(el: ParseElement): string {
  // RDFa: content attribute is the universal override.
  const content = getAttr(el, "content");
  if (content !== undefined) return content;

  const tag = el.tagName.toLowerCase();
  switch (tag) {
    case "meta":
      return ""; // unreachable: meta with no content already handled above
    case "a":
    case "link":
      return getAttr(el, "href") ?? "";
    case "img":
    case "audio":
    case "video":
    case "source":
    case "embed":
    case "iframe":
    case "track":
      return getAttr(el, "src") ?? "";
    case "object":
      return getAttr(el, "data") ?? "";
    case "data":
    case "meter":
      return getAttr(el, "value") ?? "";
    case "time": {
      const dt = getAttr(el, "datetime");
      return dt !== undefined ? dt : textContent(el).trim();
    }
    default: {
      // RDFa-specific: resource attribute is a URI reference when
      // present and no typeof is on the element.
      const resource = getAttr(el, "resource");
      if (resource !== undefined) return resource;
      return textContent(el).trim();
    }
  }
}

/**
 * Find descendants with `[property]` that belong to THIS scope.
 * Don't descend into nested `[typeof]` — they belong to the nested
 * item.
 */
function collectPropertyElements(scope: ParseElement): ParseElement[] {
  const result: ParseElement[] = [];
  function recurse(node: ParseNode) {
    if (!isElement(node)) {
      for (const c of children(node)) recurse(c);
      return;
    }
    const hasProp = hasAttr(node, "property");
    const isInnerType = hasAttr(node, "typeof") && node !== scope;

    if (hasProp) {
      result.push(node);
      // If also nested typeof, don't descend — children belong to
      // the nested item.
      if (hasAttr(node, "typeof")) return;
    } else if (isInnerType) {
      // Bare nested typeof (no property) — orphan from THIS scope.
      return;
    }
    for (const c of node.childNodes) recurse(c);
  }
  for (const c of scope.childNodes) recurse(c);
  return result;
}

export interface ExtractedRdfaItem {
  "@context"?: string;
  "@type"?: string;
  "@id"?: string;
  /** Raw typeof attribute value — diagnostic only. */
  "@itemtypeRaw"?: string;
  [propName: string]: unknown;
}

export interface RdfaExtractionResult {
  items: ExtractedRdfaItem[];
  extractionIssues: Issue[];
}

/**
 * Extract a single typeof'd item, recursing into nested typeof
 * elements for property values.
 *
 * @param currentVocab The vocab inherited from ancestors (null if
 *                     none was set). The element's own `vocab=` (if
 *                     present) overrides this.
 * @returns the extracted item, or null when the type can't be
 *          resolved (NO_VOCAB or non-schema.org vocab).
 */
function extractItem(
  el: ParseElement,
  currentVocab: string | null,
  issues: Issue[],
): ExtractedRdfaItem | null {
  const typeofVal = (getAttr(el, "typeof") ?? "").trim();
  const ownVocab = getAttr(el, "vocab");
  const vocab = ownVocab !== undefined ? ownVocab : currentVocab;

  let typeName: string | undefined;
  const itemtypeRaw = typeofVal;

  const fullUrlMatch = typeofVal.match(SCHEMA_TYPE_URL_RE);
  if (fullUrlMatch) {
    typeName = fullUrlMatch[1]!;
  } else if (typeofVal.includes(":")) {
    // CURIE — not supported in cycle 7.
    issues.push(
      invalidItemtype(
        typeofVal,
        "CURIE prefixes (schema:Product) are not supported in cycle 7. Use vocab= + bare typeof, or a fully-qualified URL.",
      ),
    );
    // continue without typeName; per-item will emit MISSING_TYPE.
  } else if (!vocab) {
    issues.push(noVocab(typeofVal));
    return null;
  } else if (!SCHEMA_VOCAB_RE.test(vocab.trim())) {
    // vocab is set but not schema.org — silently no item from this
    // subtree (the detector got us here, but this is a non-schema
    // RDFa document).
    return null;
  } else {
    // Bare name + schema.org vocab → resolved
    typeName = typeofVal;
  }

  if (typeName !== undefined && registry.types[typeName] === undefined) {
    issues.push(
      invalidItemtype(
        typeofVal,
        `Type "${typeName}" is not in the registry.`,
      ),
    );
    // keep typeName so per-item emits UNKNOWN_TYPE consistently
  }

  const item: ExtractedRdfaItem = {
    "@context": "https://schema.org",
  };
  if (typeName !== undefined) item["@type"] = typeName;
  item["@itemtypeRaw"] = itemtypeRaw;

  const resource = getAttr(el, "resource");
  if (resource !== undefined) item["@id"] = resource;

  // Collect properties.
  for (const propEl of collectPropertyElements(el)) {
    const propAttr = (getAttr(propEl, "property") ?? "").trim();
    const propNames = propAttr.split(/\s+/).filter(Boolean);
    if (propNames.length === 0) continue;

    let value: unknown;
    if (hasAttr(propEl, "typeof")) {
      // Nested item — pass our resolved vocab down
      const nested = extractItem(propEl, vocab ?? null, issues);
      value = nested ?? "";
    } else {
      value = extractElementValue(propEl);
    }
    for (const n of propNames) {
      item[n] = value;
    }
  }

  return item;
}

/**
 * Top-down walk. Tracks the current vocab as we recurse. When we
 * find an element with `[typeof]`, treat it as a top-level item
 * (don't descend further at that point — nested items are extracted
 * as part of the item's own recursion).
 */
function walkTopLevel(
  node: ParseNode,
  currentVocab: string | null,
  items: ExtractedRdfaItem[],
  issues: Issue[],
): void {
  if (!isElement(node)) {
    for (const c of children(node)) {
      walkTopLevel(c, currentVocab, items, issues);
    }
    return;
  }
  const ownVocab = getAttr(node, "vocab");
  const newVocab = ownVocab !== undefined ? ownVocab : currentVocab;

  if (hasAttr(node, "typeof")) {
    const item = extractItem(node, newVocab, issues);
    if (item !== null) items.push(item);
    return; // don't descend; nested typeofs handled by the item's own recursion
  }

  for (const c of node.childNodes) {
    walkTopLevel(c, newVocab, items, issues);
  }
}

export function extractRdfa(html: string): RdfaExtractionResult {
  const extractionIssues: Issue[] = [];

  let doc: ReturnType<typeof parse>;
  try {
    doc = parse(html);
  } catch (err) {
    extractionIssues.push(
      invalidItemtype(
        null,
        `HTML parse failure: ${err instanceof Error ? err.message : String(err)}`,
      ),
    );
    return { items: [], extractionIssues };
  }

  const items: ExtractedRdfaItem[] = [];
  walkTopLevel(doc as unknown as ParseNode, null, items, extractionIssues);

  return { items, extractionIssues };
}

File: python/src/schema_audit/_rdfa_extractor.py
"""RDFa extractor.

Ports ``core/src/utils/rdfa-extractor.ts``. Walks the HTML tree
top-down, tracking the inherited ``vocab`` as we descend. When we hit a
``[typeof]`` element, we extract it as a top-level item — descending
into nested ``[typeof]`` is handled by the item's own recursion (not
the top-level walk).

Per-element value-extraction:

- ``content=`` attribute is the universal RDFa override
- ``meta``                            → ``content``
- ``a``, ``link``                     → ``href``
- ``img``/``audio``/``video``/...     → ``src``
- ``object``                          → ``data``
- ``data``, ``meter``                 → ``value``
- ``time``                            → ``datetime`` / text
- ``[resource]`` (no ``typeof``)      → ``resource`` URI
- anything else                       → trimmed text

Cycle-10 limitations (mirror the JS cycle-7 limitations exactly):

- CURIE prefixes (``typeof="schema:Product"``) emit ``INVALID_ITEMTYPE``
  with the same message JS uses.
- Non-schema.org vocab subtree silently produces no item.

Pure: no I/O.
"""

from __future__ import annotations

import re

from ._html import (
    Element,
    TextNode,
    children,
    get_attr,
    has_attr,
    is_element,
    parse_html,
    text_content,
)
from .errors import invalid_itemtype, no_vocab
from .registry import load_registry
from .types import Issue

_REGISTRY = load_registry()

_SCHEMA_VOCAB_RE = re.compile(r"^https?://(?:www\.)?schema\.org/?$")
_SCHEMA_TYPE_URL_RE = re.compile(
    r"^https?://(?:www\.)?schema\.org/([A-Za-z][A-Za-z0-9]*)/?$"
)


def _extract_element_value(el: Element) -> str:
    """RDFa value-extraction: content= attribute is the universal override."""
    content = get_attr(el, "content")
    if content is not None:
        return content

    tag = el.tag_name.lower()
    if tag == "meta":
        return ""  # unreachable: meta with no content already handled above
    if tag in ("a", "link"):
        return get_attr(el, "href") or ""
    if tag in ("img", "audio", "video", "source", "embed", "iframe", "track"):
        return get_attr(el, "src") or ""
    if tag == "object":
        return get_attr(el, "data") or ""
    if tag in ("data", "meter"):
        return get_attr(el, "value") or ""
    if tag == "time":
        dt = get_attr(el, "datetime")
        if dt is not None:
            return dt
        return text_content(el).strip()

    # RDFa-specific: resource attribute is a URI reference when present
    # and no typeof is on the element.
    resource = get_attr(el, "resource")
    if resource is not None:
        return resource
    return text_content(el).strip()


def _collect_property_elements(scope: Element) -> list[Element]:
    """Find descendants with ``[property]`` belonging to THIS scope.

    Don't descend into nested ``[typeof]`` — those properties belong to
    the nested item.
    """
    result: list[Element] = []

    def recurse(node: Element | TextNode) -> None:
        if not is_element(node):
            for c in children(node):
                recurse(c)
            return
        assert isinstance(node, Element)

        has_prop = has_attr(node, "property")
        is_inner_type = has_attr(node, "typeof") and node is not scope

        if has_prop:
            result.append(node)
            if has_attr(node, "typeof"):
                return  # nested item; children belong to it
        elif is_inner_type:
            return  # orphan nested typeof from this scope's perspective

        for c in node.children:
            recurse(c)

    for c in scope.children:
        recurse(c)
    return result


def _extract_item(
    el: Element,
    current_vocab: str | None,
    issues: list[Issue],
) -> dict[str, object] | None:
    typeof_val = (get_attr(el, "typeof") or "").strip()
    own_vocab = get_attr(el, "vocab")
    vocab = own_vocab if own_vocab is not None else current_vocab

    type_name: str | None = None
    itemtype_raw = typeof_val

    full_url_match = _SCHEMA_TYPE_URL_RE.match(typeof_val)
    if full_url_match:
        type_name = full_url_match.group(1)
    elif ":" in typeof_val:
        issues.append(
            invalid_itemtype(
                typeof_val,
                (
                    "CURIE prefixes (schema:Product) are not supported in cycle 7. "
                    "Use vocab= + bare typeof, or a fully-qualified URL."
                ),
            )
        )
        # continue without type_name; per-item will emit MISSING_TYPE
    elif not vocab:
        issues.append(no_vocab(typeof_val))
        return None
    elif not _SCHEMA_VOCAB_RE.match(vocab.strip()):
        # vocab is set but not schema.org — silently no item from this subtree
        return None
    else:
        type_name = typeof_val

    if type_name is not None and type_name not in _REGISTRY["types"]:
        issues.append(
            invalid_itemtype(
                typeof_val,
                f'Type "{type_name}" is not in the registry.',
            )
        )
        # keep type_name so per-item emits UNKNOWN_TYPE consistently

    item: dict[str, object] = {"@context": "https://schema.org"}
    if type_name is not None:
        item["@type"] = type_name
    item["@itemtypeRaw"] = itemtype_raw

    resource = get_attr(el, "resource")
    if resource is not None:
        item["@id"] = resource

    for prop_el in _collect_property_elements(el):
        prop_attr = (get_attr(prop_el, "property") or "").strip()
        prop_names = [name for name in prop_attr.split() if name]
        if not prop_names:
            continue

        value: object
        if has_attr(prop_el, "typeof"):
            nested = _extract_item(prop_el, vocab, issues)
            value = nested if nested is not None else ""
        else:
            value = _extract_element_value(prop_el)

        for name in prop_names:
            item[name] = value

    return item


def _walk_top_level(
    node: Element | TextNode,
    current_vocab: str | None,
    items: list[dict[str, object]],
    issues: list[Issue],
) -> None:
    if not is_element(node):
        for c in children(node):
            _walk_top_level(c, current_vocab, items, issues)
        return
    assert isinstance(node, Element)

    own_vocab = get_attr(node, "vocab")
    new_vocab = own_vocab if own_vocab is not None else current_vocab

    if has_attr(node, "typeof"):
        item = _extract_item(node, new_vocab, issues)
        if item is not None:
            items.append(item)
        return  # nested typeofs handled by the item's own recursion

    for c in node.children:
        _walk_top_level(c, new_vocab, items, issues)


def extract_rdfa(html: str) -> tuple[list[dict[str, object]], list[Issue]]:
    """Parse HTML and extract RDFa items.

    Returns a 2-tuple ``(items, extraction_issues)``. ``items`` are
    JSON-LD-shaped dicts each carrying ``@context``, optional
    ``@type``, optional ``@itemtypeRaw``, and property keys.
    """
    extraction_issues: list[Issue] = []
    doc = parse_html(html)

    items: list[dict[str, object]] = []
    _walk_top_level(doc, None, items, extraction_issues)
    return items, extraction_issues
